File: packages/websearch/src/engines/wikipedia.ts
import type {
  WebSearchEngine,
  WebSearchEngineInput,
  WebSearchEngineResult,
  WebSearchResultItem,
} from "../types.js";
import { stripTags } from "./html.js";
import { httpGet } from "./http.js";
import { SearchError } from "./searchError.js";

const ENGINE_NAME = "wikipedia";

/**
 * Wikipedia / MediaWiki search API — keyless JSON, encyclopedic only.
 * Rock-solid (it never anti-bot-challenges with a descriptive User-Agent),
 * so it's the always-available tail of the fallback chain: best for factual
 * / entity queries and as a "never returns a transport error" backstop.
 *
 *   GET https://{lang}.wikipedia.org/w/api.php
 *       ?action=query&list=search&srsearch={q}&srlimit={n}&format=json
 *   → { query: { search: [{ title, pageid, snippet (html), ... }] } }
 *
 * Maps title←title, url←https://{lang}.wikipedia.org/?curid={pageid},
 * snippet←strip-tags(snippet). `language: "auto"`/unset → "en".
 *
 * @param opts.baseUrl override the API origin for tests (fixture server).
 *   In production the origin is derived from the request language.
 */
export function createWikipediaEngine(
  opts: { baseUrl?: string } = {},
): WebSearchEngine & { readonly name: string } {
  return {
    name: ENGINE_NAME,
    async search(
      input: WebSearchEngineInput,
    ): Promise<WebSearchEngineResult> {
      const lang = normalizeLang(input.language);
      const origin = opts.baseUrl ?? `https://${lang}.wikipedia.org`;
      const url = new URL(origin);
      url.pathname = joinPath(url.pathname, ["w", "api.php"]);
      const p = url.searchParams;
      p.set("action", "query");
      p.set("list", "search");
      p.set("srsearch", input.query);
      p.set("srlimit", String(input.count));
      p.set("format", "json");

      const res = await httpGet(url, input, {
        accept: "application/json",
        engine: ENGINE_NAME,
      });

      let parsed: unknown;
      try {
        parsed = JSON.parse(res.text);
      } catch (e) {
        throw new SearchError(
          "IO_ERROR",
          `wikipedia: could not parse response as JSON: ${(e as Error).message}`,
          { engine: ENGINE_NAME },
        );
      }

      const results = mapResults(parsed, lang, origin);
      return { results, backendHost: res.host, elapsedMs: res.elapsedMs };
    },
  };
}

function mapResults(
  parsed: unknown,
  _lang: string,
  origin: string,
): WebSearchResultItem[] {
  if (parsed === null || typeof parsed !== "object") return [];
  const query = (parsed as { query?: unknown }).query;
  if (query === null || typeof query !== "object") return [];
  const raw = (query as { search?: unknown }).search;
  if (!Array.isArray(raw)) return [];
  const out: WebSearchResultItem[] = [];
  for (const entry of raw) {
    if (entry === null || typeof entry !== "object") continue;
    const e = entry as {
      title?: unknown;
      pageid?: unknown;
      snippet?: unknown;
    };
    const title = typeof e.title === "string" ? e.title : "";
    if (title.length === 0) continue;
    let url = "";
    if (typeof e.pageid === "number") {
      url = `${origin.replace(/\/+$/, "")}/?curid=${e.pageid}`;
    } else {
      url = `${origin.replace(/\/+$/, "")}/wiki/${encodeURIComponent(title.replace(/ /g, "_"))}`;
    }
    const snippet = typeof e.snippet === "string" ? stripTags(e.snippet) : "";
    out.push({ title, url, snippet });
  }
  return out;
}

function normalizeLang(language: string): string {
  if (language === "" || language === "auto") return "en";
  // Take the primary subtag: "en-US" → "en".
  const primary = language.split(/[-_]/)[0] ?? "en";
  // Wikipedia language codes are lowercase ascii; reject anything weird.
  return /^[a-z]{2,3}$/.test(primary.toLowerCase())
    ? primary.toLowerCase()
    : "en";
}

function joinPath(basePath: string, segments: string[]): string {
  const trimmed = basePath.replace(/\/+$/, "");
  return `${trimmed}/${segments.join("/")}`;
}

File: crates/websearch/src/engines/wikipedia.rs
//! Wikipedia / MediaWiki search API — keyless JSON. See TS `engines/wikipedia.ts`.

use async_trait::async_trait;
use url::Url;

use super::html::strip_tags;
use super::http::http_get;
use crate::engine::{
    shared_client, SearchError, SearchErrorCode, WebSearchEngine, WebSearchEngineInput,
    WebSearchEngineResult,
};
use crate::types::{WebSearchResultItem, WebSearchTimeRange};

const ENGINE_NAME: &str = "wikipedia";

pub struct WikipediaEngine {
    client: reqwest::Client,
    /// Override the API origin for tests; production derives it from language.
    base_url: Option<String>,
}

impl WikipediaEngine {
    pub fn new() -> Self {
        Self {
            client: shared_client(),
            base_url: None,
        }
    }
    pub fn with_base_url(mut self, base: impl Into<String>) -> Self {
        self.base_url = Some(base.into());
        self
    }
}

impl Default for WikipediaEngine {
    fn default() -> Self {
        Self::new()
    }
}

#[async_trait]
impl WebSearchEngine for WikipediaEngine {
    fn name(&self) -> &str {
        ENGINE_NAME
    }

    fn engine_class(&self) -> crate::engine::EngineClass {
        crate::engine::EngineClass::Vertical
    }

    async fn search(
        &self,
        input: WebSearchEngineInput,
    ) -> Result<WebSearchEngineResult, SearchError> {
        let lang = normalize_lang(&input.language);
        let origin = self
            .base_url
            .clone()
            .unwrap_or_else(|| format!("https://{}.wikipedia.org", lang));
        let mut url = Url::parse(&origin).map_err(|_| {
            SearchError::new(
                SearchErrorCode::IoError,
                format!("invalid wikipedia origin: {}", origin),
            )
        })?;
        {
            let base_path = url.path().trim_end_matches('/').to_string();
            url.set_path(&format!("{}/w/api.php", base_path));
            url.query_pairs_mut()
                .append_pair("action", "query")
                .append_pair("list", "search")
                .append_pair("srsearch", &input.query)
                .append_pair("srlimit", &input.count.to_string())
                .append_pair("format", "json");
        }

        let res = http_get(
            &self.client,
            &url,
            &input,
            "application/json",
            ENGINE_NAME,
            &[],
        )
        .await?;

        let parsed: serde_json::Value = serde_json::from_str(&res.text).map_err(|e| {
            SearchError::new(
                SearchErrorCode::IoError,
                format!("wikipedia: could not parse response as JSON: {}", e),
            )
        })?;

        Ok(WebSearchEngineResult {
            results: map_results(&parsed, &origin),
            backend_host: res.host,
            elapsed_ms: res.elapsed_ms,
            engine: Some(ENGINE_NAME.to_string()),
            engine_class: None,
            engines: None,
            // Wikipedia search ignores recency filtering.
            time_range_applied: if input.time_range == WebSearchTimeRange::All {
                None
            } else {
                Some(false)
            },
        })
    }
}

fn map_results(parsed: &serde_json::Value, origin: &str) -> Vec<WebSearchResultItem> {
    let raw = match parsed
        .get("query")
        .and_then(|q| q.get("search"))
        .and_then(|v| v.as_array())
    {
        Some(arr) => arr,
        None => return Vec::new(),
    };
    let origin_trimmed = origin.trim_end_matches('/');
    let mut out = Vec::new();
    for entry in raw {
        let title = entry.get("title").and_then(|v| v.as_str()).unwrap_or("");
        if title.is_empty() {
            continue;
        }
        let url = match entry.get("pageid").and_then(|v| v.as_u64()) {
            Some(pageid) => format!("{}/?curid={}", origin_trimmed, pageid),
            None => format!("{}/wiki/{}", origin_trimmed, title.replace(' ', "_")),
        };
        let snippet = entry
            .get("snippet")
            .and_then(|v| v.as_str())
            .map(strip_tags)
            .unwrap_or_default();
        // `timestamp` is the article's last-edit time; surface the date
        // portion as `age` (NB: last-edit, not first-publication).
        let age = entry
            .get("timestamp")
            .and_then(|v| v.as_str())
            .and_then(iso_date);
        out.push(WebSearchResultItem {
            title: title.to_string(),
            url,
            snippet,
            age,
            score: None,
            source: None,
        });
    }
    out
}

fn normalize_lang(language: &str) -> String {
    if language.is_empty() || language == "auto" {
        return "en".to_string();
    }
    let primary = language
        .split(['-', '_'])
        .next()
        .unwrap_or("en")
        .to_ascii_lowercase();
    let ok = (2..=3).contains(&primary.len()) && primary.chars().all(|c| c.is_ascii_lowercase());
    if ok {
        primary
    } else {
        "en".to_string()
    }
}

/// Extract the YYYY-MM-DD date portion from an ISO timestamp; None if it
/// doesn't start with a date.
fn iso_date(ts: &str) -> Option<String> {
    let t = ts.trim();
    let bytes = t.as_bytes();
    if bytes.len() >= 10
        && bytes[0..4].iter().all(u8::is_ascii_digit)
        && bytes[4] == b'-'
        && bytes[5..7].iter().all(u8::is_ascii_digit)
        && bytes[7] == b'-'
        && bytes[8..10].iter().all(u8::is_ascii_digit)
    {
        Some(t[0..10].to_string())
    } else {
        None
    }
}
